zerver/views/auth.py

```diff
@@ -236,16 +251,72 @@ def finish_google_oauth2(request):
     else:
         logging.error('Google oauth2 account email not found: %s' % (body,))
         return HttpResponse(status=400)
+
     email_address = email['value']
+    if not subdomain:
+        user_profile, return_data = authenticate_remote_user(request, email_address)
+        invalid_subdomain = bool(return_data.get('invalid_subdomain'))
+        return login_or_register_remote_user(request, email_address, user_profile,
+                                             full_name, invalid_subdomain)
+
+    data = {'name': full_name, 'email': email_address, 'subdomain': subdomain}
+    # Signing the data using the SECRET_KEY of Django. The end result will
+    # be an obfuscated string which cannot be tampered with. This string
+    # will be a timestamped based signature and will not be valid after 15
+    # seconds (see log_into_subdomain # function). This string will be saved
+    # in a cookie to make it more difficult for the hacker to steal it.
+    # However, this scheme still gives the hacker a 15 seconds window to steal
+    # the cookie and use it another browser to access the account of the user
+    # but if we use secure cookies and HTTPS then even this won't happen.
+    data_signature = signing.dumps(data, salt='zerver.views.auth')
+    subdomain_login_uri = ''.join([
+        settings.EXTERNAL_URI_SCHEME,
+        subdomain,
+        '.',
+        settings.EXTERNAL_HOST,
+        reverse('zerver.views.auth.log_into_subdomain')
+    ])
+
+    domain = '.' + settings.EXTERNAL_HOST.split(':')[0]
+    response = redirect(subdomain_login_uri)
+
+    response.set_cookie('subdomain.signature', data_signature, expires=15,
+                        domain=domain)
+    return response
+
+def authenticate_remote_user(request, email_address):
+    # type: (HttpRequest, str) -> Tuple[UserProfile, Dict[str, Any]]
     return_data = {} # type: Dict[str, bool]
     user_profile = authenticate(username=email_address,
                                 realm_subdomain=get_subdomain(request),
                                 use_dummy_backend=True,
                                 return_data=return_data)
+    return user_profile, return_data
+
+def log_into_subdomain(request):
+    # type: (HttpRequest) -> HttpResponse
+    state = request.COOKIES.get('subdomain.signature')
+    if state is None:
+        logging.warning('Missing state signature')
+        return HttpResponse(status=400)
+
+    try:
+        data = signing.loads(state, salt='zerver.views.auth', max_age=15)
+    except signing.SignatureExpired:
+        # Discard state if older than 15 seconds
+        logging.warning('Signature more than 15 seconds old')
+        return HttpResponse(status=400)
```

Comment:
Where does this check is the signature is valid?  It seems like maybe you could use:
https://docs.djangoproject.com/en/1.8/ref/request-response/#django.http.HttpRequest.get_signed_cookie
to simplify things.